### ISSUE TITLE
Feature request: reusable codes?

### ISSUE THREAD

**Issue by lilmike** [warner/magic-wormhole#76]
Hi,
I thought of a feature that I could really use, but I'm not sure if it would fit in in the current wormhole archetecture. Anyway, reusable codes would be nice, in which you make a code, and then whenever you want to reuse it, perhaps for a specific friend, you can just enter it. That way, instead of having to give them the code every time, perhaps making the transfer less secure, depending on the method of transfer of the code, you could just tell them 'You have a file!' or something, and they, assumin they already know the code, could simply enter the predetermined code and get the file.
Thoughts?
-Michael.

**Comment by meejah** [comment]
I think you can already use wormhole like this -- as you can specify a code when starting either side, you can of course choose to re-use a code you've used before.

**Comment by warner** [comment]
Yeah, but re-using a short code is a really bad idea: it increases an attacker's chances of guessing, and it violates the expected properties of the code (which is that it only needs to remain secret until the transfer begins, after which it's safe to reveal).

But the idea of connecting to a previously-established destination is a great one, and I totally want to find a way to support that. The idea I've been kicking around is the use of a "pet name", which is a term from capability theory. Basically, when you do a transfer, you could choose to name the connection, and then when you want to connect to the same place later, you just refer to that name, instead of using a code at all.

You might do this by naming the recipient, like:

```
alice% wormhole send --petname=Bob FILE.pdf
-> CODE
..
bob% wormhole receive --petname=Alice
(types code)
```

then later:

```
alice% wormhole send --to=Bob OTHERFILE.pdf
bob% wormhole receive --from=Alice
```

Another option would be to use the directory as a name:

```
alice% cd ~/myfiles
alice% wormhole send FILE.pdf
-> CODE
..
bob% cd ~/Downloads
bob% wormhole receive
(types code)
```

then later:

```
alice% cd ~/myfiles
alice% wormhole send OTHERFILE.pdf
Previous connection found in /home/myfiles: re-use this connection? (y/N): y
Please tell the other side to run `wormhole receive` from the same directory as before
..
bob% cd ~/Downloads
bob% wormhole receive
Previous connection found in /home/Downloads: re-use this connection? (y/N): y
```

The down side of using the directory as a name is that it's not usually tied to a specific recipient. In the example above, Bob probably receives files into `Downloads` from lots of places.

The way this would work, internally, is with a "Wormhole Seed" (I totally just made that up, but it's conceptually sound: mature Wormholes yield seeds, and seeds can be grown into a full-sized Wormhole). Basically, each normal code-based wormhole would produce a new random code, long enough that guessing it isn't ever going to be a concern, using a long channel-id too. The code would look like `p7hyqygtzhfbczjfsws5v7m4ai-ofedxfqudnwbwhoybec2dqpy2iuyrh67ojbski5yav7ujwukpcrq`, and would be stored in `~/.wormhole/petnames` or the current directory, depending on which scheme you were using. These codes are just another derivative of the PAKE-established session key.

I'll open a new ticket to cover the Wormhole Seeds internals. Let's use this ticket to discuss the petname thing.

**Comment by warner** [comment]
Oh, the `--petname=Bob` UI I sketched out has a serious usability problem: you have to decide up-front that this transfer is something that you might want to use again in the future. Worse, both sides have to make that decision (Wormhole Seeds only work in pairs).

We could improve this a tiny bit by sending a flag: if one side uses `--petname=` to remember the Seed, the other side could prompt the user to do the same. `The other person decided to save this connection to later use: do you want to save it too? (Y/n)`. It would still require at least one side to be thinking ahead, though.

Another option is to ask at the end of the transfer. `Transfer complete: do you want to save this connection for future use? (enter petname to save under, or ENTER to skip):`. But asking that every time would make normal use kind of annoying.

A third option is to remember the Wormhole Seed for the most recent transfer in `~/.wormhole/` somewhere, and add a command that retroactively assigns a persistent name to that seed. So `Transfer complete: to save this connection for later use, please run "wormhole save-seed PETNAME" before performing your next wormhole transfer`. It's still the case that both sides have to save it, though.

A more persistent application would have an addressbook, which remembers the Seeds (or more sophisticated non-Wormhole encryption keys) under petnames, and file transfers would use the petnames exclusively. The only use of wormhole codes would be to add new entries to the addressbook. This would be super-useful (and in fact I created magic-wormhole to enable just this sort of application), but I think it's counter to the handy immediacy of the CLI tool. So I think we should continue to aim `wormhole send` as primarily a quick one-off transfer tool, but provide the persistence/repeatability as a secondary option.

**Comment by brianpburns** [comment]
meejah said:
> you can specify a code when starting either side
Is that true? I don't see that feature documented. I have been able to specify the code when receiving by including it in the wormhole receive invocation, but not on the sending side. I was thinking it would be nice to have a one-time pad on both sides and use codes sequentially, but that would require the sender to be able to specify the code when initiating the transfer.

**Comment by warner** [comment]
On the sending side you can use `--code=CODE` to bypass the "generate me a code" step. You should use a large-ish number though, to reduce the chances of accidental overlap with someone else doing the same thing, since externally-provided codes don't use centralized allocation of the channel-ids. Anything in the form of `%d-%s` will work, but I'd recommend using a number of 4 or more digits. `random.randint(1000, 9999)` is safe for the digits, but you should use os.urandom and then base16/32/64/etc conversion for the string part.

And yeah, it really needs to be documented better :)
